"Cannot compile namespaces" error with --isolatedModules and no namespaces

**TypeScript Version:**  nightly (2.3.0-dev.20170417)

**Code**

```ts
function f() {}
```

**Expected behavior:**

No error, or an error message about a non-module file in an `--isolatedModules` project.

**Actual behavior:**

`src/a.ts(1,1): error TS1208: Cannot compile namespaces when the '--isolatedModules' flag is provided.`

This occurs because we in `program.ts` `verifyCompilerOptions` to look for any source file that isn't an external module declaration, and fail on the first one, whatever it may be.
The error message should be upgraded to reflect the real reason we issue this error, and not mention namespaces.
Alternately, we could just allow files without imports and actually look for a namespace before adding this error.

@RyanCavanaugh I removed the error message's code and tested exactly what you wrote (of course with `--isolatedModules` on, the compiler outputs
    error TS2304: Cannot find name 'x'.
It errors right there, no confusion.

Here's a minimal repro case, using typescript@next (full repo at https://github.com/yang/sandbox-ts-namespaces-error):

src/a.js: (notice, no import/export keywords needed)

```js
module.exports = "hello"; // you could really put anything here, e.g. console.log('hello');
```

tsconfig.js:

```js
{
  "compilerOptions": {
    "allowJs": true,
    "isolatedModules": true,
    "noEmit": true,
    "strict": true
  },
  "include": ["src"]
}
```

Error:

```
$ ./node_modules/.bin/tsc
src/a.js:1:1 - error TS1208: Cannot compile namespaces when the '--isolatedModules' flag is provided.

1 module.exports = "hello";
  ~~~~~~


Found 1 error.
```

I too first encountered this via `create-react-app --typescript`, which instructs you to create a setupProxy.js (or setupTests.js).  CRA also sets `isolatedModules` to true.  It seems to have not been an issue until a more recent typescript version (I didn't see errors about setupProxy.js while on typescript 3.1.x, only after updating to 3.2.x or 3.3.x).

The workaround of adding .js files to your excludes works when running tsc, but - annoyingly - you still see errors from those files in editors from the TS language service (tried both VS Code and Webstorm).  (This seems to be a separate and more general issue with the TS language service not respecting excludes - but I wasn't able to find an existing issue for that.)

```
  "exclude": ["src/**/*.js"]
```

--- a/src/compiler/diagnosticMessages.json
+++ b/src/compiler/diagnosticMessages.json
@@ -659,7 +659,7 @@
         "category": "Error",
         "code": 1207
     },
-    "All files must be modules when the '--isolatedModules' flag is provided.": {
+    "'{0}' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.": {
         "category": "Error",
         "code": 1208
     },

--- a/src/compiler/program.ts
+++ b/src/compiler/program.ts
@@ -3110,7 +3110,8 @@ namespace ts {
                 const firstNonExternalModuleSourceFile = find(files, f => !isExternalModule(f) && !isSourceFileJS(f) && !f.isDeclarationFile && f.scriptKind !== ScriptKind.JSON);
                 if (firstNonExternalModuleSourceFile) {
                     const span = getErrorSpanForNode(firstNonExternalModuleSourceFile, firstNonExternalModuleSourceFile);
-                    programDiagnostics.add(createFileDiagnostic(firstNonExternalModuleSourceFile, span.start, span.length, Diagnostics.All_files_must_be_modules_when_the_isolatedModules_flag_is_provided));
+                    programDiagnostics.add(createFileDiagnostic(firstNonExternalModuleSourceFile, span.start, span.length,
+                        Diagnostics._0_cannot_be_compiled_under_isolatedModules_because_it_is_considered_a_global_script_file_Add_an_import_export_or_an_empty_export_statement_to_make_it_a_module, getBaseFileName(firstNonExternalModuleSourceFile.fileName)));
                 }
             }
             else if (firstNonAmbientExternalModuleSourceFile && languageVersion < ScriptTarget.ES2015 && options.module === ModuleKind.None) {

--- a/tests/baselines/reference/importHelpersInIsolatedModules.errors.txt
+++ b/tests/baselines/reference/importHelpersInIsolatedModules.errors.txt
@@ -1,4 +1,4 @@
-tests/cases/compiler/script.ts(1,1): error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+tests/cases/compiler/script.ts(1,1): error TS1208: 'script.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
 
 
 ==== tests/cases/compiler/external.ts (0 errors) ====
@@ -16,7 +16,7 @@ tests/cases/compiler/script.ts(1,1): error TS1208: All files must be modules whe
 ==== tests/cases/compiler/script.ts (1 errors) ====
     class A { }
     ~~~~~
-!!! error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+!!! error TS1208: 'script.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
     class B extends A { }
     
     declare var dec: any;

--- a/tests/baselines/reference/isolatedModulesNoExternalModule.errors.txt
+++ b/tests/baselines/reference/isolatedModulesNoExternalModule.errors.txt
@@ -1,7 +1,7 @@
-tests/cases/compiler/file1.ts(1,1): error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+tests/cases/compiler/file1.ts(1,1): error TS1208: 'file1.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
 
 
 ==== tests/cases/compiler/file1.ts (1 errors) ====
     var x;
     ~~~
-!!! error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+!!! error TS1208: 'file1.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.

--- a/tests/baselines/reference/isolatedModulesOut.errors.txt
+++ b/tests/baselines/reference/isolatedModulesOut.errors.txt
@@ -1,6 +1,6 @@
 error TS5053: Option 'out' cannot be specified with option 'isolatedModules'.
 tests/cases/compiler/file1.ts(1,1): error TS6131: Cannot compile modules using option 'out' unless the '--module' flag is 'amd' or 'system'.
-tests/cases/compiler/file2.ts(1,1): error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+tests/cases/compiler/file2.ts(1,1): error TS1208: 'file2.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
 
 
 !!! error TS5053: Option 'out' cannot be specified with option 'isolatedModules'.
@@ -11,4 +11,4 @@ tests/cases/compiler/file2.ts(1,1): error TS1208: All files must be modules when
 ==== tests/cases/compiler/file2.ts (1 errors) ====
     var y;
     ~~~
-!!! error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+!!! error TS1208: 'file2.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.

--- a/tests/baselines/reference/isolatedModulesPlainFile-AMD.errors.txt
+++ b/tests/baselines/reference/isolatedModulesPlainFile-AMD.errors.txt
@@ -1,9 +1,9 @@
-tests/cases/compiler/isolatedModulesPlainFile-AMD.ts(1,1): error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+tests/cases/compiler/isolatedModulesPlainFile-AMD.ts(1,1): error TS1208: 'isolatedModulesPlainFile-AMD.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
 
 
 ==== tests/cases/compiler/isolatedModulesPlainFile-AMD.ts (1 errors) ====
     declare function run(a: number): void;
     ~~~~~~~
-!!! error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+!!! error TS1208: 'isolatedModulesPlainFile-AMD.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
     run(1);
     

--- a/tests/baselines/reference/isolatedModulesPlainFile-CommonJS.errors.txt
+++ b/tests/baselines/reference/isolatedModulesPlainFile-CommonJS.errors.txt
@@ -1,9 +1,9 @@
-tests/cases/compiler/isolatedModulesPlainFile-CommonJS.ts(1,1): error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+tests/cases/compiler/isolatedModulesPlainFile-CommonJS.ts(1,1): error TS1208: 'isolatedModulesPlainFile-CommonJS.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
 
 
 ==== tests/cases/compiler/isolatedModulesPlainFile-CommonJS.ts (1 errors) ====
     declare function run(a: number): void;
     ~~~~~~~
-!!! error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+!!! error TS1208: 'isolatedModulesPlainFile-CommonJS.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
     run(1);
     

--- a/tests/baselines/reference/isolatedModulesPlainFile-ES6.errors.txt
+++ b/tests/baselines/reference/isolatedModulesPlainFile-ES6.errors.txt
@@ -1,9 +1,9 @@
-tests/cases/compiler/isolatedModulesPlainFile-ES6.ts(1,1): error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+tests/cases/compiler/isolatedModulesPlainFile-ES6.ts(1,1): error TS1208: 'isolatedModulesPlainFile-ES6.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
 
 
 ==== tests/cases/compiler/isolatedModulesPlainFile-ES6.ts (1 errors) ====
     declare function run(a: number): void;
     ~~~~~~~
-!!! error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+!!! error TS1208: 'isolatedModulesPlainFile-ES6.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
     run(1);
     

--- a/tests/baselines/reference/isolatedModulesPlainFile-System.errors.txt
+++ b/tests/baselines/reference/isolatedModulesPlainFile-System.errors.txt
@@ -1,9 +1,9 @@
-tests/cases/compiler/isolatedModulesPlainFile-System.ts(1,1): error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+tests/cases/compiler/isolatedModulesPlainFile-System.ts(1,1): error TS1208: 'isolatedModulesPlainFile-System.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
 
 
 ==== tests/cases/compiler/isolatedModulesPlainFile-System.ts (1 errors) ====
     declare function run(a: number): void;
     ~~~~~~~
-!!! error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+!!! error TS1208: 'isolatedModulesPlainFile-System.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
     run(1);
     

--- a/tests/baselines/reference/isolatedModulesPlainFile-UMD.errors.txt
+++ b/tests/baselines/reference/isolatedModulesPlainFile-UMD.errors.txt
@@ -1,9 +1,9 @@
-tests/cases/compiler/isolatedModulesPlainFile-UMD.ts(1,1): error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+tests/cases/compiler/isolatedModulesPlainFile-UMD.ts(1,1): error TS1208: 'isolatedModulesPlainFile-UMD.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
 
 
 ==== tests/cases/compiler/isolatedModulesPlainFile-UMD.ts (1 errors) ====
     declare function run(a: number): void;
     ~~~~~~~
-!!! error TS1208: All files must be modules when the '--isolatedModules' flag is provided.
+!!! error TS1208: 'isolatedModulesPlainFile-UMD.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
     run(1);
     

--- a/tests/baselines/reference/tscWatch/emit/emit-for-configured-projects/should-always-return-the-file-itself-if-'--isolatedModules'-is-specified.js
+++ b/tests/baselines/reference/tscWatch/emit/emit-for-configured-projects/should-always-return-the-file-itself-if-'--isolatedModules'-is-specified.js
@@ -37,7 +37,7 @@ Output::
 [[90m12:00:23 AM[0m] Starting compilation in watch mode...
 
 
-[96ma/b/globalFile3.ts[0m:[93m1[0m:[93m1[0m - [91merror[0m[90m TS1208: [0mAll files must be modules when the '--isolatedModules' flag is provided.
+[96ma/b/globalFile3.ts[0m:[93m1[0m:[93m1[0m - [91merror[0m[90m TS1208: [0m'globalFile3.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
 
 [7m1[0m interface GlobalFoo { age: number }
 [7m [0m [91m~~~~~~~~~[0m
@@ -136,7 +136,7 @@ Output::
 [[90m12:00:38 AM[0m] File change detected. Starting incremental compilation...
 
 
-[96ma/b/globalFile3.ts[0m:[93m1[0m:[93m1[0m - [91merror[0m[90m TS1208: [0mAll files must be modules when the '--isolatedModules' flag is provided.
+[96ma/b/globalFile3.ts[0m:[93m1[0m:[93m1[0m - [91merror[0m[90m TS1208: [0m'globalFile3.ts' cannot be compiled under '--isolatedModules' because it is considered a global script file. Add an import, export, or an empty 'export {}' statement to make it a module.
 
 [7m1[0m interface GlobalFoo { age: number }
 [7m [0m [91m~~~~~~~~~[0m

